src/wp-includes/fonts/class-wp-font-face.php

```diff
@@ -0,0 +1,400 @@
+ *
+ * @since 6.4.0
+ */
+class WP_Font_Face {
+
+	/**
+	 * The font-face property defaults.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @var string[]
+	 */
+	private $font_face_property_defaults = array(
+		'font-family'  => '',
+		'font-style'   => 'normal',
+		'font-weight'  => '400',
+		'font-display' => 'fallback',
+	);
+
+	/**
+	 * Valid font-face property names.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @var string[]
+	 */
+	private $valid_font_face_properties = array(
+		'ascent-override',
+		'descent-override',
+		'font-display',
+		'font-family',
+		'font-stretch',
+		'font-style',
+		'font-weight',
+		'font-variant',
+		'font-feature-settings',
+		'font-variation-settings',
+		'line-gap-override',
+		'size-adjust',
+		'src',
+		'unicode-range',
+	);
+
+	/**
+	 * Valid font-display values.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @var string[]
+	 */
+	private $valid_font_display = array( 'auto', 'block', 'fallback', 'swap', 'optional' );
+
+	/**
+	 * Array of font-face style tag's attribute(s)
+	 * where the key is the attribute name and the
+	 * value is its value.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @var string[]
+	 */
+	private $style_tag_attrs = array();
+
+	/**
+	 * Creates and initializes an instance of WP_Font_Face.
+	 *
+	 * @since 6.4.0
+	 */
+	public function __construct() {
+		if (
+			function_exists( 'is_admin' ) && ! is_admin()
+			&&
+			function_exists( 'current_theme_supports' ) && ! current_theme_supports( 'html5', 'style' )
+		) {
+			$this->style_tag_attrs = array( 'type' => 'text/css' );
+		}
+	}
+
+	/**
+	 * Generates and prints the `@font-face` styles for the given fonts.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param array[][] $fonts Optional. The font-families and their font variations.
+	 *                         See {@see wp_print_font_faces()} for the supported fields.
+	 *                         Default empty array.
+	 * }
+	 */
+	public function generate_and_print( array $fonts ) {
+		$fonts = $this->validate_fonts( $fonts );
+
+		// Bail out if there are no fonts are given to process.
+		if ( empty( $fonts ) ) {
+			return;
+		}
+
+		printf(
+			$this->get_style_element(),
+			$this->get_css( $fonts )
+		);
+	}
+
+	/**
+	 * Validates each of the font-face properties.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param array $fonts The fonts to valid.
+	 * @return array Prepared font-faces organized by provider and font-family.
+	 */
+	private function validate_fonts( array $fonts ) {
+		$validated_fonts = array();
+
+		foreach ( $fonts as $font_faces ) {
+			foreach ( $font_faces as $font_face ) {
+				$font_face = $this->validate_font_face_properties( $font_face );
+				// Skip if failed validation.
+				if ( false === $font_face ) {
+					continue;
+				}
+
+				$validated_fonts[] = $font_face;
+			}
+		}
+
+		return $validated_fonts;
+	}
+
+	/**
+	 * Validates each font-face property.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param array $font_face Font face properties to validate.
+	 * @return false|array Validated font-face on success. Else, false.
```

Comment:
```suggestion
	 * @return array|false Validated font-face on success. Otherwise false.
```
- Swapped return types. You pessimist! :joy:
- "Otherwise false" is Core's usual phrasing.
- Also, since this validates both properties and values, and returns the validated font face on success rather than valid properties, does it make more sense to call this `validate_font_face()`?

Reply:
Good catch on the swapped return types. Whoopsie. 

>"Otherwise false" is Core's usual phrasing.

Hmm, there's a mixture of phrasing in Core:
- `. Otherwise, false`
- [the success]`, false on failure.`
- `, or false on failure.`

I think I'll change it to the last one.

Reply:
Done in https://github.com/WordPress/wordpress-develop/pull/5051/commits/4a1d4c2b55ee54cb8563ad06ed9237ce5f73ab6a ✅

Reply:
>Also, since this validates both properties and values, and returns the validated font face on success rather than valid properties, does it make more sense to call this validate_font_face()?

I think that method name would be confusing due to:

* The font-face array contains a collection of `@font-face` property / value pairings that relate directly to the CSS. 
*  The method above it called `validate_fonts()`.

`properties` here is a verbose attempt to infer property / value pairs.

Other more accurate names could be:

* `validate_font_face_property_value_pairs()`
* `validate_font_face_declarations()` <= [the proper name for a property / value pair](https://developer.mozilla.org/en-US/docs/Web/CSS/Syntax#css_declarations)

@costdev Do you find `properties` to be confusing? Is either of the above more understandable?

Reply:
Thanks for the clarification!

I think `properties` while verbose was also a bit vague as the code validates properties and values.

`validate_font_face_declarations()` makes more sense to me as it encompasses both, so there's no surprises when seeing the internal code.

Reply:
Done in commit https://github.com/WordPress/wordpress-develop/pull/5051/commits/c8e4c03546418aae76f62ba1dc07dd2193b847e7 ✅ 

Also updated the DocBlock to ensure "declaration" is well understood.